src/commands_upstream.c

```diff
@@ -0,0 +1,546 @@
+                                mptcpd_token_t token)
+{
+        /*
+          MPTCP connection token is not currently needed by upstream
+          kernel to remove network address.
+        */
+        if (token != 0)
+                l_warn("remove_addr: MPTCP connection token "
+                       "is ignored.");
+
+        /*
+          Payload (nested):
+                  Local address ID
+         */
+
+        size_t const payload_size = MPTCPD_NLA_ALIGN(address_id);
+
+        struct l_genl_msg *const msg =
+                l_genl_msg_new_sized(MPTCP_PM_CMD_DEL_ADDR, payload_size);
+
+        bool const appended =
+                l_genl_msg_enter_nested(msg, MPTCP_PM_ATTR_ADDR)
+                && l_genl_msg_append_attr(
+                        msg,
+                        MPTCP_PM_ADDR_ATTR_ID,
+                        sizeof(address_id),
+                        &address_id)
+                && l_genl_msg_leave_nested(msg);
+
+        if (!appended) {
+                l_genl_msg_unref(msg);
+
+                return ENOMEM;
+        }
+
+        return l_genl_family_send(pm->family,
+                                  msg,
+                                  mptcpd_family_send_callback,
+                                  NULL, /* user data */
+                                  NULL  /* destroy */)
+                == 0;
+}
+
+static int upstream_get_addr(struct mptcpd_pm *pm,
+                             mptcpd_aid_t address_id,
+                             struct mptcpd_addr_info **addr)
+{
+        /*
+          Payload (nested):
+              Local address ID
+         */
+
+        size_t const payload_size = MPTCPD_NLA_ALIGN(address_id);
+
+        struct l_genl_msg *const msg =
+                l_genl_msg_new_sized(MPTCP_PM_CMD_GET_ADDR,
+                                     payload_size);
+
+        bool const appended =
+                l_genl_msg_enter_nested(msg, MPTCP_PM_ATTR_ADDR)
+                && l_genl_msg_append_attr(
+                        msg,
+                        MPTCP_PM_ADDR_ATTR_ID,
+                        sizeof(address_id),
+                        &address_id)
+                && l_genl_msg_leave_nested(msg);
+
+        if (!appended) {
+                l_genl_msg_unref(msg);
+
+                return ENOMEM;
+        }
+
+        struct dumped_addrs dump = { .addrs = NULL };
+
+        /**
+         * @todo This is assumed to be a synchronous call.  Verify.
+         *       If not, we'll likely need to manage the addr memory
+         *       differently.
+         */
+        int const request_id =
+                l_genl_family_send(pm->family,
+                                   msg,
+                                   get_addr_callback,
+                                   &dump, /* user data */
+                                   NULL  /* destroy */);
+
+        if (request_id != 0) {
+                *addr = dump.addrs;
+                assert(dump.len == 1);
```

Comment:
This `assert()` is triggered when running the `test-commands` unit test.  It looks like `l_genl_family_send()` doesn't block waiting for the full reply from the kernel, i.e. it's an async call.  A similar problem should occur when calling `dump_addr()` and `get_limits()`.  I'll modify these three functions to accept a user-supplied callback function argument that will be called once the reply containing the requested data from the kernel is available.

Reply:
Commit 3203d10 implements the user callbacks.